core/domain/certificate_assessment_domain_test.py

```
@@ -250,3 +252,392 @@ def test_from_dict_then_to_dict_matches_original_dict(self) -> None:
+    def test_validate_rejects_invalid_attempt_index(self) -> None:
+        attempt = self._get_sample_attempt()
+        attempt.attempt_index = 0
+        with self.assertRaisesRegex(
+            Exception, 'attempt_index must be a positive integer'
+        ):
+            attempt.validate()
+
+    def test_validate_rejects_empty_attempt_data(self) -> None:
+        attempt = self._get_sample_attempt()
+        attempt.attempt_data = {}
+        with self.assertRaisesRegex(
```

Ruff (0.15.20)

[warning] Use `pytest.raises` instead of unittest-style `assertRaisesRegex`

Replace `assertRaisesRegex` with `pytest.raises`

(PT027)